src/NzbDrone.Core/Messaging/Commands/CommandService.cs

```diff
@@ -0,0 +1,193 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using NLog;
+using NzbDrone.Common;
+using NzbDrone.Common.Cache;
+using NzbDrone.Common.EnsureThat;
+using NzbDrone.Common.Serializer;
+using NzbDrone.Core.Jobs;
+using NzbDrone.Core.Lifecycle;
+using NzbDrone.Core.Messaging.Events;
+
+namespace NzbDrone.Core.Messaging.Commands
+{
+    public interface ICommandService
+    {
+        CommandModel PublishCommand<TCommand>(TCommand command) where TCommand : Command;
+        CommandModel PublishCommand(string commandName);
+        void PublishScheduledTasks(List<ScheduledTask> commands);
+        CommandModel Pop();
+        CommandModel Get(int id);
+        List<CommandModel> GetStarted(); 
+        void SetMessage(CommandModel command, string message);
+        void Completed(CommandModel command);
+        void Failed(CommandModel command, Exception e);
+    }
+
+    public class CommandService : ICommandService, IHandle<ApplicationStartedEvent>
+    {
+        private readonly ICommandRepository _repo;
+        private readonly IServiceFactory _serviceFactory;
+        private readonly IEventAggregator _eventAggregator;
+        private readonly Logger _logger;
+
+        private ICached<string> _messageCache; 
+
+        private static readonly object Mutex = new object();
+
+        public CommandService(ICommandRepository repo, 
+                              IServiceFactory serviceFactory,
+                              IEventAggregator eventAggregator,
+                              ICacheManager cacheManager,
+                              Logger logger)
+        {
+            _repo = repo;
+            _serviceFactory = serviceFactory;
+            _eventAggregator = eventAggregator;
+            _logger = logger;
+
+            _messageCache = cacheManager.GetCache<string>(GetType());
+        }
+
+        public CommandModel PublishCommand<TCommand>(TCommand command) where TCommand : Command
+        {
+            Ensure.That(command, () => command).IsNotNull();
+
+            _logger.Trace("Publishing {0}", command.GetType().Name);
+
+            lock (Mutex)
+            {
+                var existingCommands = _repo.FindCommands(command.Name).Where(c => c.Status == CommandStatus.Queued ||
```

Comment:
This should run in the DB, otherwise it'll get really heavy getting a list of all `CheckForCompletedDownloads` and filtering it in memory.